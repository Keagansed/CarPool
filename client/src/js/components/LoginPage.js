import { observer } from "mobx-react";
import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';

//import { getFromStorage } from '../utils/localStorage.js';

import fbIcon from "./../../css/images/fb_icon.png";
import googleIcon from "./../../css/images/google_icon.png";

@observer class Login extends Component {
    constructor(props){
        super(props);
        // ========= Properties ===========
        this.state ={
          token:'',
        }; 
    }

    //========= Render Component ===========
    componentWillMount(){
        // const obj = getFromStorage('sessionKey');
        // if(obj && obj.token){
        //     //verify token
        //     const { token } = obj;
        //     fetch('/api/account/verify?token='+token)
        //     .then(res => res.json())
        //     .then(json => {
        //         if(json.success){
        //             this.props.store.setToken(token);
        //             this.props.store.setLoggedIn(true);
        //         }
        //     })
        // }
    }
    componentDidMount(){}

    updateLoginEmailValue = event =>
    {
        this.props.store.lEmail = event.target.value;
    }

    updateLoginPasswordValue = event =>
    {
        this.props.store.lPassword = event.target.value;
    }

    handleLogin = event => {
        event.preventDefault();
        this.props.store.authenticate(this.state.email, this.state.password);
    }

    render() {

        const { loggedIn } = this.props.store;
    
        if(!loggedIn){
            return(
                <div className="text-center">
                    <div className="container">
                        <div className="row bigTopMargin">
                            <div className="col-md-12">
                                <form className="">
                                    <div className="form-group">
                                        <input onChange={this.updateLoginEmailValue} type="email" className="form-control roundCorners bg-info border-1_5 border-primary w-75 mx-auto text-white" placeholder="Email" required="required" name="txtEmail" id="txtEmail"/> 
                                    </div>
                                    <div className="form-group">
                                        <input onChange={this.updateLoginPasswordValue} type="password" className="form-control roundCorners bg-info border-1_5 border-primary w-75 mx-auto text-white" placeholder="Password" required="required" name="txtPassword" id="txtPassword"/> 
                                    </div>
                                    <p className="my-2" id="dividerOR">
                                        <strike>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</strike> &nbsp; <b className="small">Forgot Password?</b> &nbsp;
                                        <strike>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</strike>
                                    </p>
                                    <button onClick={this.handleLogin} type="submit" className="btn btn-primary roundCorners w-75 text-secondary my-3 font-weight-medium">
                                        <b>Login</b>
                                    </button>
                                </form>
                            </div>
                        </div>
                    </div>

                    <div className="text-center">
                        <div className="container">
                            <div className="row">
                                <div className="col-md-12">
                                    <p id="dividerOR">
                                        <b className="small font-weight-medium">OR</b>
                                    </p>
                                </div>
                            </div>
                        </div>
                    </div>

                    <div className="p-0">
                        <div className="text-center container">
                            <div className="row text-center">
                                <div className="col-2-pr-5 col-sm-2 col-md-6 col-lg-2 col-xl-2 ml-auto">
                                    <img className="img-fluid d-block mx-auto" src={fbIcon} id="btnFacebookLogin" alt="fbLogin"/> 
                                </div>
                                <div className="col-2-pl-5 col-sm-2 col-md-6 col-lg-2 col-xl-2 mr-auto">
                                    <img className="img-fluid d-block mx-auto" src={googleIcon} id="btnGoogleLogin" alt="googleLogin"/> 
                                </div>
                            </div>
                        </div>
                    </div>

                </div>
            );
        }else{

            return(               
                 <Redirect to={{
                     pathname: "/HomePage", 
                 }}/>
             ); 
        }
    }

}

export default Login;