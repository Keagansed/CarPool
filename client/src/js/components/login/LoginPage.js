import { observer } from "mobx-react";
import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';

import "./../../../css/components/Login.css"

import logo  from "./../../../css/images/logo.png";
import { getFromStorage } from './../../utils/localStorage.js';
import Modal from './modal';

@observer class Login extends Component {
    constructor(props){
        super(props);
        // ========= Properties ===========
        this.state ={
          token:'',
        }; 
    }

    //========= Render Component ===========
    componentWillMount(){
        const obj = getFromStorage('sessionKey');
        if(obj && obj.token){
            //verify token
            const { token } = obj;
            fetch('/api/account/verify?token='+token)
            .then(res => res.json())
            .then(json => {
                if(json.success){
                    this.props.store.setToken(token);
                    this.props.store.setLoggedIn(true);
                }
            })
        }
    }
    componentDidMount(){}

    updateLoginEmailValue = event =>
    {
        this.props.store.lEmail = event.target.value;
    }

    updateLoginPasswordValue = event =>
    {
        this.props.store.lPassword = event.target.value;
    }

    handleLogin = event => {
        event.preventDefault();
        this.props.store.authenticate(this.state.email, this.state.password);
    }

    openModal = event =>{
        event.preventDefault();
        console.log("link clicked");
        this.modal.open();
    }

    render() {

        const { loggedIn } = this.props.store;
    
        if(!loggedIn){
            return(
                <div className="vertical-center bg-purple">
                    <div className="container-fluid">
                        <div className="row">
                                <img className="img-fluid d-block mx-auto mbottom-2rem" src={logo} id="imgLogo" alt="carpool_logo"/> 
                        </div>
                        <form>
                            <div className="row">
                                <input onChange={this.updateLoginEmailValue} type="email" className="form-control mx-auto width-15rem brad-2rem mbottom-1rem" placeholder="Email" required="required" name="Email" id="inputEmail"/> 
                            </div>
                            <div className="row">
                                <input onChange={this.updateLoginPasswordValue} type="password" className="form-control mx-auto width-15rem brad-2rem mbottom-1rem" placeholder="Password" required="required" name="Password" id="inputPassword"/> 
                            </div>
                            <div className="row">
                                <button onClick={this.handleLogin} type="submit" className="btn btn-primary mx-auto width-15rem brad-2rem mbottom-1rem bg-aqua txt-purple fw-bold" id="btnLogin">
                                    Login
                                </button>
                            </div>
                        </form>
                        <div className="row">
                            <a data-toggle="modal" data-target="#myModal" className="mx-auto txt-white fw-100" id="linkForgotPassword">
                                Forgot Password?
                            </a>
                        </div>
                    </div>
                    <Modal />
                </div>
            );
        }else{

            return(               
                 <Redirect to={{
                     pathname: "/HomePage", 
                 }}/>
             ); 
        }
    }

}

export default Login;