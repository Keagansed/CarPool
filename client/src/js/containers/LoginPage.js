// File Type: Component

import { observer } from "mobx-react";
import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';

import Modal from '../components/login/PasswordModal';
import { getFromStorage } from '../utils/localStorage.js';

import logo  from "../../css/images/logo.png";

const util = require('./../utils/idCheck');
const display = {
    display: 'block'
};
const hide = {
    display: 'none'
};

/*
* Purpose: Validate whether all of the fields are valid - true if there are errors
*/
function validate(email, password) {
    return {
        email: !util.ValidateEmail(email),
        password: password.length === 0,
    };
}


/*
* Purpose: Login page where the user enters login details to proceed to the HomePage
*/
@observer class Login extends Component {
    constructor(props) {
        super(props);
        this.toggle = this.toggle.bind(this);
        
        this.state ={
            toggle: false,
            token:'',
            email: '',
            password: '',

            touched: {
                email: false,
                password: false,
            },
        }; 
    }

    /*
     * Purpose: toggles the visibility of the component.  
     */
    toggle(event) {
        this.setState(prevState => ({
            toggle: !prevState.toggle
        }));
    }

    /*
    * Purpose: Check to see whether the user is already logged in
    */
    componentWillMount() {
        const obj = getFromStorage('sessionKey');
        if(obj && obj.token) {
            const { token } = obj;

            fetch('/api/account/verify?token='+token)
            .then(res => res.json())
            .then(json => {
                if(json.success) {
                    this.props.store.setToken(token);
                    this.props.store.setLoggedIn(true);
                }
            })
        }
    }

    /*
    * Purpose: Sets the 'store.lEmail' variable to senders current value
    */
    updateLoginEmailValue = (event) => {
        this.setState({ email: event.target.value });
        this.props.store.lEmail = event.target.value;
    }

    /*
    * Purpose: Sets the 'store.lPassword' variable to senders current value
    */
    updateLoginPasswordValue = (event) => {
        this.setState({ password: event.target.value });
        this.props.store.lPassword = event.target.value;
    }

    /*
    * Purpose: Calls the store.authenticate() function
    */
    handleLogin = (event) => {
        event.preventDefault();
        if (!this.canBeSubmitted()) {
            return;
        }
        this.props.store.authenticate(this.state.email, this.state.password);
        //TODO: this toggle should only be called if authentication was unsuccessful
        this.toggle();
    }

    /*
    * Purpose: Check whether all fields have been entered correctly
    */
    canBeSubmitted() {
        const errors = validate(this.state.email, this.state.password);
        const isDisabled = Object.keys(errors).some(x => errors[x]);
        return !isDisabled;
    }

    /*
    * Purpose: Give fields that have been entered incorrectly red borders
    */
    handleBlur = (field) => (evt) => {
        this.setState({
            touched: { ...this.state.touched, [field]: true },
        });
    }

    render() {
        /*
        * Purpose: Only give fields red borders if the user has changed/access them
        * and they are still not valid.
        */
        const shouldMarkError = (field) => {
            const hasError = errors[field];
            const shouldShow = this.state.touched[field];
            return hasError ? shouldShow : false;
        };

        const { loggedIn } = this.props.store;
        const errors = validate(this.state.email, this.state.password);
        const isDisabled = Object.keys(errors).some(x => errors[x]);

        var modal = [];
        modal.push(
            // Modal
            <div key="0" className="modal" tabIndex="-1" role="dialog" id="myModal" style={this.state.toggle ? display : hide}>
                <div className="modal-dialog" role="document">
                    <div className="modal-content">
                        <div className="modal-header">
                            <h5 className="modal-title">Login Failed</h5>
                            <button type="button" className="close" onClick={this.toggle} aria-label="Close">
                            <span aria-hidden="true">&times;</span>
                            </button>
                        </div>
                        <div className="modal-body">
                            You have entered an incorrect email or password.
                        </div>
                    </div>
                </div>
            </div>
        );
    
        if(!loggedIn) {
            return(
                <div className="vertical-center bg-purple">
                    <div className="container-fluid">
                        <div className="row">
                            <img 
                                className="img-fluid d-block mx-auto mbottom-2rem" 
                                src={logo} 
                                id="imgLogo" 
                                alt="carpool_logo"
                            /> 
                        </div>
                        <form>
                            <div className="row">
                                <input 
                                    onChange={this.updateLoginEmailValue} 
                                    type="email" 
                                    className={(shouldMarkError('email') ? "error" : "") + " form-control mx-auto width-15rem brad-2rem mbottom-1rem"}
                                    onBlur={this.handleBlur('email')}
                                    placeholder="Email" 
                                    id="inputEmail"
                                    value={this.state.email}
                                /> 
                            </div>
                            <div className="row">
                                <input 
                                    onChange={this.updateLoginPasswordValue} 
                                    type="password" 
                                    className={(shouldMarkError('password') ? "error" : "") + " form-control mx-auto width-15rem brad-2rem mbottom-1rem"}
                                    onBlur={this.handleBlur('password')}
                                    placeholder="Password"
                                    id="inputPassword"
                                    value={this.state.password}
                                /> 
                            </div>
                            <div className="row">
                                <button 
                                    onClick={this.handleLogin} 
                                    type="submit" 
                                    className="btn btn-primary mx-auto width-15rem brad-2rem mbottom-1rem bg-aqua txt-purple fw-bold" 
                                    id="btnLogin"
                                    disabled={isDisabled}
                                >
                                    Login
                                </button>
                            </div>
                        </form>
                        <div className="row">
                            {/* forgot password modal */}
                            <Modal />
                        </div>
                        {/* login failed modal */}
                        {modal}
                    </div>
                </div>
            );
        }else {

            return(               
                 <Redirect to={{
                     pathname: "/HomePage", 
                 }}/>
             ); 
        }
    }

}

export default Login;