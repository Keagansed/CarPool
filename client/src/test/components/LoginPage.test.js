import React from 'react';
import LoginPage from '../../js/containers/LoginPage';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

class MockStore {
    token = null; 
    loggedIn = false; 
    registered = false; 
    lemail = ''; 
    lpassword = '';
    sFName = '';
    sLName = '';
    semail = '';
    sPass1 = '';
    sPass2 = '';
    sId = '';

    setToken = (token) => {
        this.token = token;
    };

    setLoggedIn = (success) => {
        this.loggedIn = success;
        this.lemail = '';
        this.lpassword = '';
    };

    setRegistered = (val) => {
        this.registered = val;
    };

    signup = () => {
        
        if(this.sPass1 !== this.sPass2) {
            alert("Passwords do not match");
        }else{
            alert("Successfully signed up");
            this.setRegistered(true);
        }
    };

    authenticate = () => {
        this.setRegistered(false);
        this.setToken("testtoken123");
        this.setLoggedIn(true);
    }

    logout = () => {
        this.setLoggedIn(false);
        this.setToken('');
    }

}

describe('Login Page Component', () => {
    let container, instance;
    const mockStore = new MockStore();
    const loginPage = new LoginPage();

    beforeAll(() => {
        container = shallow(<LoginPage store={mockStore} />);
        instance = container.instance();
    });

    it('captures snapshot', () => {
        const renderedValue = renderer.create(<LoginPage store={mockStore} />).toJSON();
        expect(renderedValue).toMatchSnapshot();
    });

    it('renders correctly', () => {
        expect(container.length).toEqual(1)
    });

    it('updates login email value correctly', () => {
        const event = {target: {value: "test@email.com"}};
        expect(instance.state.email).toEqual('');
        instance.updateLoginEmailValue(event);
        expect(instance.state.email).toEqual("test@email.com");
    })

    it('updates login password value correctly', () => {
        const event = {target: {value: "1234"}};
        expect(instance.state.password).toEqual('');
        instance.updateLoginPasswordValue(event);
        expect(instance.state.password).toEqual("1234");
    })

    it('checks correct state variables correctly', () => {
        expect(instance.canBeSubmitted()).toEqual(true);
    })

    it('check state variables correctly with incorrect email', () => {
        instance.state.email = "invalidemail";
        expect(instance.canBeSubmitted()).toEqual(false);
    })

    it('check state variables correctly with incorrect password', () => {
        instance.state.email = "test@email.com";
        instance.state.password = "";
        expect(instance.canBeSubmitted()).toEqual(false);
    })
});