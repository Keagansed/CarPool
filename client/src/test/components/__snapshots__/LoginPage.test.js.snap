// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page Component captures snapshot 1`] = `
<div
  className="vertical-center bg-purple"
>
  <div
    className="container-fluid"
  >
    <div
      className="row"
    >
      <img
        alt="carpool_logo"
        className="img-fluid d-block mx-auto mbottom-2rem"
        id="imgLogo"
        src="logo.png"
      />
    </div>
    <form>
      <div
        className="row"
      >
        <input
          className=" form-control mx-auto width-15rem brad-2rem mbottom-1rem"
          id="inputEmail"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Email"
          type="email"
          value=""
        />
      </div>
      <div
        className="row"
      >
        <input
          className=" form-control mx-auto width-15rem brad-2rem mbottom-1rem"
          id="inputPassword"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
      <div
        className="row"
      >
        <button
          className="btn btn-primary mx-auto width-15rem brad-2rem mbottom-1rem bg-aqua txt-purple fw-bold"
          disabled={true}
          id="btnLogin"
          onClick={[Function]}
          type="submit"
        >
          Login
        </button>
      </div>
    </form>
    <div
      className="row"
    >
      <div
        className="mx-auto"
      >
        <a
          className="txt-white fw-100"
          onClick={[Function]}
        >
          Forgot Password?
        </a>
        <div
          className="modal"
          id="myModal"
          role="dialog"
          style={
            Object {
              "display": "none",
            }
          }
          tabIndex="-1"
        >
          <div
            className="modal-dialog"
            role="document"
          >
            <div
              className="modal-content"
            >
              <div
                className="modal-header"
              >
                <h5
                  className="modal-title"
                >
                  Forgot Password?
                </h5>
                <button
                  aria-label="Close"
                  className="close"
                  onClick={[Function]}
                  type="button"
                >
                  <span
                    aria-hidden="true"
                  >
                    ×
                  </span>
                </button>
              </div>
              <div
                className="modal-body"
              >
                <form>
                  <div
                    className="row"
                  >
                    <input
                      className="form-control mx-auto width-15rem brad-2rem mbottom-1rem"
                      id="inputForgotEmail"
                      name="Email"
                      placeholder="Email"
                      required="required"
                      type="email"
                    />
                  </div>
                  <div
                    className="row"
                  >
                    <button
                      className="btn btn-primary mx-auto width-15rem brad-2rem mbottom-1rem bg-aqua txt-purple fw-bold"
                      id="btnForgotPassword"
                      onClick={undefined}
                      type="submit"
                    >
                      Send Me My Password
                    </button>
                  </div>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="modal"
      id="myModal"
      role="dialog"
      style={
        Object {
          "display": "none",
        }
      }
      tabIndex="-1"
    >
      <div
        className="modal-dialog"
        role="document"
      >
        <div
          className="modal-content"
        >
          <div
            className="modal-header"
          >
            <h5
              className="modal-title"
            >
              Login Failed
            </h5>
            <button
              aria-label="Close"
              className="close"
              onClick={[Function]}
              type="button"
            >
              <span
                aria-hidden="true"
              >
                ×
              </span>
            </button>
          </div>
          <div
            className="modal-body"
          >
            You have entered an incorrect email or password.
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
